Add  credentials validation and required field notice for Payfast in sandbox environment.

iamdharmesh opened the pull request:
### All Submissions:

<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
This PR adds credentials validation in the Payfast payment settings when the settings are saved and displays a corresponding notice. It also makes changes to show a required credentials field notice when the credentials are blank, and it prevents Payfast from appearing on the front-end checkout page. Previously, this required notice was only implemented for the production environment. This PR ensures it works for the sandbox environment as well.

Closes #216 

### Steps to test the changes in this Pull Request:
1. Remove all Payfast credentials from the settings and save the settings.
1. Verify that a notice with the required fields error is visible.
1. Try saving invalid credentials and verify that the merchant is prompted with an invalid credentials error.

### Changelog entry
> Add - Credentials validation and required field notice for PayFast in the sandbox environment.

iamdharmesh committed:
Display required field notice on sandbox env and validate payfast credentials on save settings.

iamdharmesh committed:
Added E2E tests.

iamdharmesh committed:
Fix formatting issue.

ankitguptaindia committed:
Merge branch 'trunk' into fix/216

ankitguptaindia commented:
QA Status: Needs feedback
------


 Thanks, @iamdharmesh, for the fixes! I’ve identified scenarios that need to be addressed.

**Fields Required Tests:**

- When all fields are empty, it shows the correct message. ✅
- When any single field is empty and others are filled, it displays a notice. ✅
- When fields are empty and settings are saved, the PayFast option should not appear on the checkout page. ✅

**Credentials Verification:**

- When invalid credentials are entered, it should display an error. ✅
- When invalid credentials are entered, the PayFast option should not be visible on the checkout page. ⚠️ 


**_When the admin adds invalid credentials, a warning notice is displayed in the backend, but the PayFast payment option still appears on the checkout page. Is it possible to hide the PayFast option on the frontend when the credentials are invalid?_** 

https://github.com/user-attachments/assets/dba50e2c-ddc0-4441-b46f-18b35c30f226

iamdharmesh committed:
Make credentials validation strict and prevent showing payfast in list if invalid credentials are entered.

iamdharmesh commented:
Thanks for testing this @ankitguptaindia,

> When the admin adds invalid credentials, a warning notice is displayed in the backend, but the PayFast payment option still appears on the checkout page. Is it possible to hide the PayFast option on the frontend when the credentials are invalid?

Sure, I had added normal credentials validation to show only notice. But now I have updated it to strict validation where notice will be keep showing and payment gateway will not display in the front-end until merchant enter the correct credentials. Could you please help to check once now.

Thank you.

ankitguptaindia commented:
**QA/Test Report-** ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.6.2
* Theme active on store: Twenty Twenty-Four Version: 1.2
* WooCommerce - Version 9.3.3
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 129
* OS: macOS Sonoma 14.6.1 

</Details>


**Test Results -** 

**Fields Required Tests:**

- When all fields are empty, it shows the correct message. ✅
- When any single field is empty and others are filled, it displays a notice. ✅
- When fields are empty and settings are saved, the PayFast option should not appear on the checkout page. ✅

**Credentials Verification:**

- When invalid credentials are entered, it should display an error. ✅
- When invalid credentials are entered, the PayFast option should not be visible on the checkout page. ✅

ankitguptaindia commented:
**Regression+Smoke Test Report-** ✅ 
------------

**Testing Environment -** 


<Details>

* WordPress: 6.3.2
* Theme: Twenty Twenty-Four Version: 1.2
* WooCommerce - Version 9.3.3
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 129
* OS: macOS Sonoma 14.5


</Details>

Tested with Archive File created via woorelease command. 


- Composer version - v2.3.5
- npm version - v10.1.0
- node version - v20.7.0

#### Please note that this plugin has been tested with the build created by the specified versions ☝🏼 of Composer, Node, and NPM. 



**Status-**  Working as expected. Ready to merge 🚀
